pact4s/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/provider/LeoProvider.scala

```diff
@@ -83,38 +62,16 @@ class LeoProvider extends AnyFlatSpec with BeforeAndAfterAll with PactVerifier {
       mockAppService,
       mockRuntimeV2Service,
       mockAdminService,
-      MockResourcesService,
+      mockResourcesService,
       mockUserInfoDirectives,
       mockContentSecurityPolicyConfig,
       refererConfig
     )
 
   private val providerStatesHandler: StateManagementFunction = StateManagementFunction {
-    case ProviderState(States.AppExists, _) =>
-      when(mockAppService.getApp(any[UserInfo], any[CloudContext.Gcp], AppName(anyString()))(any[Ask[IO, AppContext]]))
-        .thenReturn(IO {
-          GetAppResponse(
-            None,
-            AppName("exampleApp"),
-            CloudContext.Gcp(GoogleProject("exampleProject")),
-            RegionName("exampleRegion"),
-            KubernetesRuntimeConfig(NumNodes(8), MachineTypeName("exampleMachine"), autoscalingEnabled = true),
-            List.empty[AppError],
-            AppStatus.Unspecified,
-            Map.empty[KubernetesSerializableName.ServiceName, URL],
-            Some(DiskName("exampleDiskName")),
-            Map.empty[String, String],
-            AuditInfo(WorkbenchEmail(""), Instant.now(), None, Instant.now()),
-            AppType.CromwellRunnerApp,
-            ChartName(""),
-            None,
-            Map.empty[String, String],
-            false,
-            None
-          )
-        })
-    case _ =>
-      loggerIO.debug("other state")
+    RuntimeStateManager.handler(mockRuntimeService).orElse(AppStateManager.handler(mockAppService)).orElse { case _ =>
```

Comment:
huh... a little confused why there are multiple `orElse` chained here, seems like a bit of a code smell but I'm not super familiar with what this code is doing enough to know for sure. I'd recommend trying to change this to a function that has a match statement if possible, psuedocode below
```
determiningThing match {
case 1 =>
case 2 =>
case _ => 
```

Reply:
I think chaining partial functions encourages separation of duties and re-usability of code (like a runtime state manager and another state manager for the rest of the Leo app), the effect is going to be the same as using a `case match` block.

Reply:
the downside is its not super clear why one is taking precedence over another (which is what my main point of confusion is). Don't feel super strong about switching the syntax though since semantically it would be the same, a comment could suffice

Reply:
If the states are mutually exclusive then I think there is no precedence one way over the other. @evrii would you add a comment to your code for clarity sake?

Reply:
Thanks @jdcanas and @ichengchang .

I added a comment for clarification.

@jdcanas I would prefer a case statement here too, but I have not been able to figure out a way to handle that.

```
  // This function composes a large switch statement based on partial functions from
  // multiple state mangers (each with their own state that they handle).
  private val providerStatesHandler: StateManagementFunction = StateManagementFunction {
    RuntimeStateManager.handler(mockRuntimeService).orElse(AppStateManager.handler(mockAppService)).orElse { case _ =>
      loggerIO.debug("Whoops: other state")
    }
  }
```

The way this works is that a given state is passed in (say X in this case).

This function checks RuntimeStateManager and it finds that it has cases for A,B, and C but no X

It then proceeds to AppStateManager where is finds states X, Y, and Z. So it executes the code in the X case.

I could see this working like this, but I haven't been able to get it to work:
case (A|B|C):
   Call RuntimeStateManager
       case A:
           ...
      case B:
           ...
      case C:
           ...
case (X|Y|Z):
    Call AppStateManager
       case X:
           ...
      case Y:
           ...
      case Z:
           ...